src/rotation_types.jl

```diff
@@ -0,0 +1,192 @@
+            uses the solution with ||SpQuat|| <= 1 
+
+""" ->
+immutable SpQuat{T <: Real}
+    x::T
+    y::T
+    z::T
+end
+
+
+# convert to a known one
+convert_rotation{T <: Real}(::Type{Quaternion{T}}, spq::SpQuat{T}) = spquat_to_quat(spq)
+convert_rotation{T <: Real}(::Type{SpQuat{T}}, q::Quaternion{T}) = quat_to_spquat(q)
+
+# define its interaction with other angle representations
+convert_rotation{T <: Real}(::Type{SpQuat{T}}, R::RotMatrix{T}) = quat_to_spquat(convert_rotation(Quaternion{T}, R))
+convert_rotation{T,U}(::Type{T}, spq::SpQuat{U}) = convert_rotation(T, spquat_to_quat(promote_eltype(spq, T))) # this has template overload problems without the U????
+
+# default parameters
+default_params{T <: SpQuat}(::Type{T}) = (@DefaultElType(), )
+
+# define an inverse for the SPQuat since its trivial.  Inverse in the sense of the corresponding inverse rotation 
+inv(X::SpQuat) = SpQuat(-X.x, -X.y, -X.z) 
+
+# define multiplication for SpQuat corresponding to combining rotations
+*(lhs::SpQuat, rhs::SpQuat) = SpQuat(Quaternion(lhs) * Quaternion(rhs))
+
+# pull the rotation angle as well
+rot_angle(X::SpQuat) = rot_angle(Quaternion(X))
+
+
+
+###################################################
+# Arbitrary axis rotation representations
+# (limited support for now)
+###################################################
+
+
+immutable AngleAxis{T <: Real}  
+    theta::T
+    axis_x::T
+    axis_y::T
+    axis_z::T
+end
+
+# define its interaction with other angle representations
+convert_rotation{T <: Real}(::Type{Quaternion{T}}, aa::AngleAxis{T}) = arbaxis_to_quat(aa)
+convert_rotation{T <: Real}(::Type{AngleAxis{T}}, q::Quaternion{T}) = quat_to_arbaxis(q)
+
+# go via the quaternion representation
+convert_rotation{T <: Real}(::Type{AngleAxis{T}}, R::RotMatrix{T}) = quat_to_arbaxis(convert_rotation(Quaternion{T}, R))
+convert_rotation{T,U}(::Type{T}, aa::AngleAxis{U}) = convert_rotation(T, arbaxis_to_quat(promote_eltype(aa, eltype(T))))  # this has template overload problems without the U????
+
+# default parameters
+default_params{T <: AngleAxis}(::Type{T}) = (@DefaultElType(), )
+
+# accessors
+rot_angle(aa::AngleAxis) = aa.theta - floor((aa.theta+pi) / (2*pi)) * 2*pi  # named to match the quaternion equivilent
+axis(aa::AngleAxis) = Vec(aa.axis_x, aa.axis_y, aa.axis_z)
+
+# construct with funny inputs
+call(::Type{AngleAxis}, theta::Real, x::Real, y::Real, z::Real) = add_params(AngleAxis, promote_type(typeof(theta), typeof(x), typeof(y), typeof(z)))(theta,x,y,z)
+
+
+
+
+###################################################
+# Euler Angles
+###################################################
+
+# this file defines the types:
+# EulerAngles       - for tait bryan ordering
+# ProperEulerAngles - for proper Euler ordering
+include("euler_types.jl") 
+
+default_params{T <: EulerAngles}(::Type{T}) = (@DefaultEulerOrder(), @DefaultElType()) 
```

Comment:
Why not use a Union{} type? Or create an abstract type for EulerAngles and ProperEulerAngles?